File: internal/skills/skill/validation.go
// Package skill loads, validates, and imports portable Agent Skill bundles.
package skill

import (
	"fmt"
	"path/filepath"
	"slices"
	"strings"

	"github.com/danieljustus/symaira-brain/internal/skills/variant"
)

func ValidateSkillName(name string) error {
	if strings.TrimSpace(name) == "" {
		return fmt.Errorf("skill name is required")
	}
	if len(name) > MaxNameLength {
		return fmt.Errorf("skill name %q exceeds maximum length of %d characters", name, MaxNameLength)
	}
	if !skillNamePattern.MatchString(name) {
		return fmt.Errorf("skill name %q must be a single lowercase alphanumeric-dash segment without consecutive, leading, or trailing hyphens", name)
	}
	return nil
}

// Validate returns non-fatal validation issues for a loaded bundle.
func Validate(bundle *Bundle) []Issue {
	if bundle == nil {
		return []Issue{{Code: "bundle_required", Severity: "error", Message: "bundle is nil"}}
	}
	var issues []Issue
	name := bundle.Frontmatter.Name
	if strings.TrimSpace(name) == "" {
		issues = append(issues, Issue{Code: "name_required", Severity: "error", Message: "frontmatter name is required", Path: "SKILL.md"})
	} else {
		if len(name) > MaxNameLength {
			issues = append(issues, Issue{Code: "name_too_long", Severity: "error", Message: fmt.Sprintf("name exceeds maximum length of %d characters (actual: %d)", MaxNameLength, len(name)), Path: "SKILL.md"})
		}
		if !skillNamePattern.MatchString(name) {
			issues = append(issues, Issue{Code: "name_format", Severity: "error", Message: "name must use lowercase letters, numbers, and single dashes (no consecutive, leading, or trailing dashes)", Path: "SKILL.md"})
		}
		// The agentskills specification requires the skill name to match the
		// parent directory name, which is how the library layout stores
		// bundles (library/<name>).
		if filepath.Base(bundle.Root) != name {
			issues = append(issues, Issue{Code: "name_dir_mismatch", Severity: "error", Message: fmt.Sprintf("frontmatter name %q must match the parent directory name %q", name, filepath.Base(bundle.Root)), Path: "SKILL.md"})
		}
	}
	if strings.TrimSpace(bundle.Frontmatter.Description) == "" {
		issues = append(issues, Issue{Code: "description_required", Severity: "error", Message: "frontmatter description is required", Path: "SKILL.md"})
	} else if len(bundle.Frontmatter.Description) > MaxDescriptionLength {
		issues = append(issues, Issue{Code: "description_too_long", Severity: "error", Message: fmt.Sprintf("frontmatter description exceeds maximum length of %d characters (actual: %d)", MaxDescriptionLength, len(bundle.Frontmatter.Description)), Path: "SKILL.md"})
	}
	if strings.TrimSpace(bundle.Frontmatter.Category) == "" {
		issues = append(issues, Issue{Code: "category_required", Severity: "warning", Message: "frontmatter category is required; reuse an existing category when possible", Path: "SKILL.md"})
	}
	if strings.TrimSpace(bundle.Body) == "" {
		issues = append(issues, Issue{Code: "body_required", Severity: "error", Message: "SKILL.md body is empty", Path: "SKILL.md"})
	} else if len(bundle.Body) > MaxBodyLength {
		issues = append(issues, Issue{Code: "body_too_long", Severity: "warning", Message: fmt.Sprintf("SKILL.md body exceeds maximum length of %d characters (actual: %d)", MaxBodyLength, len(bundle.Body)), Path: "SKILL.md"})
	}
	for _, res := range bundle.Resources {
		if res.Executable {
			issues = append(issues, Issue{Code: "resource_executable", Severity: "warning", Message: "resource file is executable; install strips the executable bit unless --allow-executable (or the manifest setting) is set", Path: res.Path})
		}
		if res.Size > MaxResourceSize {
			issues = append(issues, Issue{Code: "resource_too_large", Severity: "error", Message: fmt.Sprintf("resource exceeds maximum size of %d bytes (actual: %d)", MaxResourceSize, res.Size), Path: res.Path})
		}
	}
	targets := make([]string, 0, len(bundle.Manifest.Targets))
	for target := range bundle.Manifest.Targets {
		targets = append(targets, target)
	}
	slices.Sort(targets)
	for _, target := range targets {
		cfg := bundle.Manifest.Targets[target]
		if !cfg.Enabled {
			continue
		}
		for _, rel := range []string{cfg.Prepend, cfg.Append} {
			if rel == "" {
				continue
			}
			if err := safeRelativeFile(bundle, rel); err != nil {
				issues = append(issues, Issue{Code: "overlay_reference_missing", Severity: "error", Message: err.Error(), Path: "symskills.toml:" + target})
			}
		}
	}
	issues = append(issues, validateVariants(bundle)...)
	return issues
}

// renderBlockingCodes lists the validation codes that make a render refuse
// rather than emit output. They share one property: the rendered result
// would silently misrepresent the source — a traversing overlay reference, a
// malformed marker that would ship as literal HTML, an override addressing a
// block that does not exist, or a placeholder that cannot resolve. Every
// other error-severity code is reported by `validate` and still renders.
var renderBlockingCodes = map[string]bool{
	"overlay_reference_missing":     true,
	variant.CodeMarkerMalformed:     true,
	variant.CodeBlockIDInvalid:      true,
	variant.CodeBlockNested:         true,
	variant.CodeBlockUnclosed:       true,
	variant.CodeBlockUnmatchedClose: true,
	variant.CodeBlockCloseMismatch:  true,
	variant.CodeBlockDuplicateID:    true,
	variant.CodeTargetListEmpty:     true,
	variant.CodeOverrideUnknown:     true,
	variant.CodeTermUnknown:         true,
	variant.CodeTermNameInvalid:     true,
	variant.CodeTermDefaultRequired: true,
}

// IsRenderBlocking reports whether an error-severity validation code must
// stop a render instead of only being reported.
func IsRenderBlocking(code string) bool { return renderBlockingCodes[code] }

// variantIssues converts variant problems into validation issues for one
// path. lineOffset shifts a problem's line to its real position in the file,
// which is non-zero for SKILL.md, whose body starts below the frontmatter.
func variantIssues(path string, lineOffset int, problems []variant.Problem) []Issue {
	issues := make([]Issue, 0, len(problems))
	for _, problem := range problems {
		message := problem.Message
		if problem.Line > 0 {
			message = fmt.Sprintf("line %d: %s", problem.Line+lineOffset, problem.Message)
		}
		issues = append(issues, Issue{Code: problem.Code, Severity: problem.Severity, Message: message, Path: path})
	}
	return issues
}

// validateVariants checks the harness-variant constructs across SKILL.md and
// every markdown reference: marker structure, bundle-wide unique block ids,
// resolvable terms, and overlay overrides that address a block the canonical
// source actually defines. The last check is what keeps an overlay a delta
// keyed to the source instead of a second, drifting copy of it.
func validateVariants(bundle *Bundle) []Issue {
	var issues []Issue
	known := knownTargets()

	paths := make([]string, 0, len(bundle.Markdown)+1)
	for path := range bundle.Markdown {
		paths = append(paths, path)
	}
	slices.Sort(paths)
	paths = append([]string{"SKILL.md"}, paths...)

	// Coupling is only meaningful for a skill that targets more than one
	// harness. A skill deliberately scoped to a single enabled target may
	// name it freely.
	checkCoupling := len(known) > 0 && enabledTargetCount(bundle) != 1

	var sourceIDs []string
	definedIn := map[string]string{}
	for _, path := range paths {
		text := bundle.Body
		lineOffset := bundle.BodyLineOffset
		if path != "SKILL.md" {
			text = bundle.Markdown[path]
			lineOffset = 0
		}
		if checkCoupling {
			for _, mention := range variant.FindMentions(text, known) {
				issues = append(issues, Issue{
					Code:     variant.CodeHarnessCoupling,
					Severity: variant.SeverityWarning,
					Message: fmt.Sprintf(
						"line %d: names harness %q outside any symskills:only region; every other target renders this text too — scope it with a region, move the differing part into a {{term:...}}, or disable the other targets in symskills.toml",
						mention.Line+lineOffset, mention.Name),
					Path: path,
				})
			}
		}
		scan, problems := variant.ScanText(text)
		issues = append(issues, variantIssues(path, lineOffset, problems)...)
		issues = append(issues, variantIssues(path, lineOffset, variant.CheckRegionTargets(scan.Regions, known))...)
		for _, id := range scan.BlockIDs {
			if prev, dup := definedIn[id]; dup {
				issues = append(issues, Issue{
					Code:     variant.CodeBlockDuplicateID,
					Severity: variant.SeverityError,
					Message:  fmt.Sprintf("block id %q is already defined in %s; ids address a block across the whole skill and must be unique", id, prev),
					Path:     path,
				})
				continue
			}
			definedIn[id] = path
			sourceIDs = append(sourceIDs, id)
		}
		for _, name := range scan.Terms {
			if _, ok := bundle.Manifest.Terms[name]; !ok {
				issues = append(issues, Issue{
					Code:     variant.CodeTermUnknown,
					Severity: variant.SeverityError,
					Message:  fmt.Sprintf("term %q is referenced but not defined in [terms]", name),
					Path:     path,
				})
			}
		}
	}

	overrideIDs := map[string][]string{}
	dirs := make([]string, 0, len(bundle.BlockOverrides))
	for dir := range bundle.BlockOverrides {
		dirs = append(dirs, dir)
	}
	slices.Sort(dirs)
	for _, dir := range dirs {
		for id := range bundle.BlockOverrides[dir] {
			overrideIDs[dir] = append(overrideIDs[dir], id)
		}
		if cfg, ok := bundle.Manifest.Targets[dir]; ok && !cfg.Enabled {
			issues = append(issues, Issue{
				Code:     variant.CodeOverrideUnused,
				Severity: variant.SeverityWarning,
				Message:  fmt.Sprintf("target %q is disabled but ships block overrides; they are never rendered", dir),
				Path:     "overlays/" + dir + "/" + variant.BlocksDir,
			})
		}
	}
	issues = append(issues, variantIssues("overlays", 0, variant.CheckOverrides(sourceIDs, overrideIDs))...)
	issues = append(issues, variantIssues("symskills.toml", 0, variant.CheckTerms(bundle.Manifest.Terms, known))...)
	return issues
}

// enabledTargetCount counts the targets a manifest explicitly enables. A
// skill with no manifest returns 0, which renders for every target.
func enabledTargetCount(bundle *Bundle) int {
	count := 0
	for _, cfg := range bundle.Manifest.Targets {
		if cfg.Enabled {
			count++
		}
	}
	return count
}

func safeRelativeFile(bundle *Bundle, rel string) error {
	if filepath.IsAbs(rel) {
		return fmt.Errorf("overlay reference %q must be relative", rel)
	}
	clean := filepath.Clean(rel)
	if strings.HasPrefix(clean, ".."+string(filepath.Separator)) || clean == ".." {
		return fmt.Errorf("overlay reference %q escapes skill root", rel)
	}
	if _, err := ReadBundleBytes(bundle, filepath.ToSlash(clean), MaxInputSize); err != nil {
		return fmt.Errorf("overlay reference %q: %w", rel, err)
	}
	return nil
}

File: rust/symbrain-skills/src/validation.rs
//! Bundle validation and render-blocking classification.

use std::collections::BTreeMap;
use std::path::Path;

use crate::load::read_bundle_text;
use crate::model::{
    Bundle, Issue, MAX_BODY_LENGTH, MAX_DESCRIPTION_LENGTH, MAX_NAME_LENGTH, MAX_RESOURCE_SIZE,
};
use crate::variant;

fn valid_slug(value: &str, separator: char) -> bool {
    !value.is_empty()
        && value
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == separator)
        && !value.starts_with(separator)
        && !value.ends_with(separator)
        && !value.contains("--")
}

fn normalize_issue(code: &str, severity: &str, message: impl Into<String>, path: &str) -> Issue {
    Issue {
        code: code.into(),
        severity: severity.into(),
        message: message.into(),
        path: path.into(),
    }
}
fn blocking(code: &str) -> bool {
    matches!(
        code,
        "overlay_reference_missing"
            | variant::CODE_MARKER_MALFORMED
            | variant::CODE_BLOCK_ID_INVALID
            | variant::CODE_BLOCK_NESTED
            | variant::CODE_BLOCK_UNCLOSED
            | variant::CODE_BLOCK_UNMATCHED_CLOSE
            | variant::CODE_BLOCK_CLOSE_MISMATCH
            | variant::CODE_BLOCK_DUPLICATE_ID
            | variant::CODE_TARGET_LIST_EMPTY
            | variant::CODE_OVERRIDE_UNKNOWN
            | variant::CODE_TERM_UNKNOWN
            | variant::CODE_TERM_NAME_INVALID
            | variant::CODE_TERM_DEFAULT_REQUIRED
    )
}

/// Returns whether an issue code refuses rendering.
#[must_use]
pub fn is_render_blocking(code: &str) -> bool {
    blocking(code)
}

/// Built-in harness target names used by the public convenience validator.
pub const DEFAULT_TARGETS: &[&str] = &[
    "opencode",
    "claude",
    "codex",
    "hermes",
    "antigravity",
    "openclaw",
];

/// Validates a bundle against the built-in target registry.
///
/// Call [`validate_with_targets`] when a caller has registered custom targets.
#[must_use]
pub fn validate(bundle: &Bundle) -> Vec<Issue> {
    let targets = DEFAULT_TARGETS
        .iter()
        .map(|target| (*target).to_owned())
        .collect::<Vec<_>>();
    validate_with_targets(bundle, &targets)
}

/// Validates a bundle, including known-target and cross-file variant checks.
#[must_use]
#[allow(
    clippy::too_many_lines,
    reason = "validation keeps issue ordering identical to the Go oracle"
)]
pub fn validate_with_targets(bundle: &Bundle, known_targets: &[String]) -> Vec<Issue> {
    let mut issues = Vec::new();
    let fm = &bundle.frontmatter;
    if fm.name.trim().is_empty() {
        issues.push(normalize_issue(
            "name_required",
            "error",
            "frontmatter name is required",
            "SKILL.md",
        ));
    } else {
        if fm.name.len() > MAX_NAME_LENGTH {
            issues.push(normalize_issue(
                "name_too_long",
                "error",
                format!(
                    "name exceeds maximum length of {MAX_NAME_LENGTH} characters (actual: {})",
                    fm.name.len()
                ),
                "SKILL.md",
            ));
        }
        if !valid_slug(&fm.name, '-') {
            issues.push(normalize_issue("name_format", "error", "name must use lowercase letters, numbers, and single dashes (no consecutive, leading, or trailing dashes)", "SKILL.md"));
        }
        if bundle.root.file_name().and_then(|name| name.to_str()) != Some(fm.name.as_str()) {
            issues.push(normalize_issue(
                "name_dir_mismatch",
                "error",
                format!(
                    "frontmatter name {:?} must match the parent directory name {:?}",
                    fm.name,
                    bundle
                        .root
                        .file_name()
                        .and_then(|name| name.to_str())
                        .unwrap_or("")
                ),
                "SKILL.md",
            ));
        }
    }
    if fm.description.trim().is_empty() {
        issues.push(normalize_issue(
            "description_required",
            "error",
            "frontmatter description is required",
            "SKILL.md",
        ));
    } else if fm.description.len() > MAX_DESCRIPTION_LENGTH {
        issues.push(normalize_issue("description_too_long", "error", format!("frontmatter description exceeds maximum length of {MAX_DESCRIPTION_LENGTH} characters (actual: {})", fm.description.len()), "SKILL.md"));
    }
    if fm.category.trim().is_empty() {
        issues.push(normalize_issue(
            "category_required",
            "warning",
            "frontmatter category is required; reuse an existing category when possible",
            "SKILL.md",
        ));
    }
    if bundle.body.trim().is_empty() {
        issues.push(normalize_issue(
            "body_required",
            "error",
            "SKILL.md body is empty",
            "SKILL.md",
        ));
    } else if bundle.body.len() > MAX_BODY_LENGTH {
        issues.push(normalize_issue(
            "body_too_long",
            "warning",
            format!(
                "SKILL.md body exceeds maximum length of {MAX_BODY_LENGTH} characters (actual: {})",
                bundle.body.len()
            ),
            "SKILL.md",
        ));
    }
    for resource in &bundle.resources {
        if resource.executable {
            issues.push(normalize_issue("resource_executable", "warning", "resource file is executable; install strips the executable bit unless --allow-executable (or the manifest setting) is set", &resource.path));
        }
        if resource.size > MAX_RESOURCE_SIZE {
            issues.push(normalize_issue(
                "resource_too_large",
                "warning",
                format!(
                    "resource exceeds maximum size of {MAX_RESOURCE_SIZE} bytes (actual: {})",
                    resource.size
                ),
                &resource.path,
            ));
        }
    }
    for (target, config) in &bundle.manifest.targets {
        if !config.enabled {
            continue;
        }
        for reference in [&config.prepend, &config.append] {
            if reference.is_empty() {
                continue;
            }
            if let Err(message) = safe_relative_file(bundle, reference) {
                issues.push(normalize_issue(
                    "overlay_reference_missing",
                    "error",
                    message,
                    &format!("symskills.toml:{target}"),
                ));
            }
        }
    }
    let mut paths = bundle.markdown.keys().cloned().collect::<Vec<_>>();
    paths.sort();
    paths.insert(0, "SKILL.md".into());
    let single_target = bundle
        .manifest
        .targets
        .values()
        .filter(|config| config.enabled)
        .count()
        == 1;
    let check_coupling = !known_targets.is_empty() && !single_target;
    let mut definitions = BTreeMap::<String, String>::new();
    let mut source_ids = Vec::new();
    for path in paths {
        let markdown_text;
        let (text, offset) = if path == "SKILL.md" {
            (&bundle.body as &str, bundle.body_line_offset)
        } else {
            markdown_text = String::from_utf8_lossy(&bundle.markdown[&path]);
            (markdown_text.as_ref(), 0)
        };
        if check_coupling {
            for mention in variant::find_mentions(text, known_targets) {
                issues.push(normalize_issue(variant::CODE_HARNESS_COUPLING, "warning", format!("line {}: names harness {:?} outside any symskills:only region; every other target renders this text too", mention.line + offset, mention.name), &path));
            }
        }
        let (scan, problems) = variant::scan_text(text);
        for problem in problems
            .into_iter()
            .chain(variant::check_region_targets(&scan.regions, known_targets))
        {
            issues.push(normalize_issue(
                &problem.code,
                &problem.severity,
                if problem.line > 0 {
                    format!("line {}: {}", problem.line + offset, problem.message)
                } else {
                    problem.message
                },
                &path,
            ));
        }
        for id in scan.block_ids {
            if let Some(previous) = definitions.get(&id) {
                issues.push(normalize_issue(variant::CODE_BLOCK_DUPLICATE_ID, "error", format!("block id {id:?} is already defined in {previous}; ids address a block across the whole skill and must be unique"), &path));
            } else {
                definitions.insert(id.clone(), path.clone());
                source_ids.push(id);
            }
        }
        for term in scan.terms {
            if !bundle.manifest.terms.contains_key(&term) {
                issues.push(normalize_issue(
                    variant::CODE_TERM_UNKNOWN,
                    "error",
                    format!("term {term:?} is referenced but not defined in [terms]"),
                    &path,
                ));
            }
        }
    }
    let mut overrides = BTreeMap::new();
    for (target, blocks) in &bundle.block_overrides {
        overrides.insert(target.clone(), blocks.keys().cloned().collect::<Vec<_>>());
        if bundle
            .manifest
            .targets
            .get(target)
            .is_some_and(|config| !config.enabled)
        {
            issues.push(normalize_issue(variant::CODE_OVERRIDE_UNUSED, "warning", format!("target {target:?} is disabled but ships block overrides; they are never rendered"), &format!("overlays/{target}/blocks")));
        }
    }
    for problem in variant::check_overrides(&source_ids, &overrides) {
        issues.push(normalize_issue(
            &problem.code,
            &problem.severity,
            problem.message,
            "overlays",
        ));
    }
    for problem in variant::check_terms(&bundle.manifest.terms, known_targets) {
        issues.push(normalize_issue(
            &problem.code,
            &problem.severity,
            problem.message,
            "symskills.toml",
        ));
    }
    issues
}

fn safe_relative_file(bundle: &Bundle, reference: &str) -> Result<(), String> {
    let path = Path::new(reference);
    if path.is_absolute() {
        return Err(format!("overlay reference {reference:?} must be relative"));
    }
    if path
        .components()
        .any(|component| matches!(component, std::path::Component::ParentDir))
    {
        return Err(format!(
            "overlay reference {reference:?} escapes skill root"
        ));
    }
    read_bundle_text(
        bundle,
        path,
        &format!("overlay reference {reference:?}"),
        crate::model::MAX_INPUT_SIZE,
    )
    .map(|_| ())
    .map_err(|error| {
        if error.0.contains("No such file or directory") || error.0.contains("not found") {
            // Keep the user-facing diagnostic byte-for-byte with Go's os.Root
            // error on the supported Unix platforms.
            format!(
                "overlay reference {reference:?}: read {reference}: openat {reference}: no such file or directory"
            )
        } else {
            error.to_string()
        }
    })
}
